### PR TITLE
chore(ci): Run the LTE integ tests on bazel-built services in new workflow

### LINKED ISSUE #11614: Run (python) services built with bazel on magma-dev vm and precommit integration tests
### Run (python) services built with bazel on magma-dev vm and precommit integration tests
Currently the services running in the magma-dev vm are built with make. It is desirable to further test the bazel-built python services (see #9878) in a more realistic setting - such as the magma-dev vm - where the bazel-built services are able to call each other.

#### Setting up the bazel-built services (manually)

- Build the already bazelified services (listed in #9878) in `/magma` on the magma-dev vm:
  - `bazel build orc8r/gateway/python/magma/eventd`
  - `bazel build orc8r/gateway/python/magma/magmad`
  - `bazel build lte/gateway/python/magma/mobilityd`
  - `bazel build lte/gateway/python/magma/redirectd`
  - `bazel build lte/gateway/python/magma/policydb`

- Go to `/etc/systemd/system/` on the magma-dev vm and modify the corresponding `magma@magmad.service `, `magma@eventd.service `, ... service files:
  - `#ExecStart=/usr/bin/env python3 -m magma.eventd.main`
      `ExecStart=/usr/bin/python3 /home/vagrant/magma/bazel-bin/orc8r/gateway/python/magma/eventd/eventd`
  - `#ExecStart=/usr/bin/env python3 -m magma.%i.main`
    `ExecStart=/usr/bin/python3 /home/vagrant/magma/bazel-bin/orc8r/gateway/python/magma/magmad/magmad`
  - `#ExecStart=/usr/bin/env python3 -m magma.mobilityd.main`
    `ExecStart=/usr/bin/python3 /home/vagrant/magma/bazel-bin/lte/gateway/python/magma/mobilityd/mobilityd`
  - `#ExecStart=/usr/bin/env python3 -m magma.redirectd.main`
   `ExecStart=/usr/bin/python3 /home/vagrant/magma/bazel-bin/lte/gateway/python/magma/redirectd/redirectd`
  - The policydb service is a special case as the .service file is created from the template file `magma@.service`. 
    - Create a copy of `magma@.service` and name it `magma@policydb.service`.
    - `#ExecStart=/usr/bin/env python3 -m magma.%i.main`
       `ExecStart=/usr/bin/python3 /home/vagrant/magma/bazel-bin/lte/gateway/python/magma/policydb/policydb`
- Load modifications with `sudo systemctl daemon-reload`
- Stop all magma services `sudo systemctl stop magma@*`
- Start the magmad service `sudo systemctl start magma@magmad.service`
- Verify that the bazel services have been started `sudo systemctl status magma@*`
  - The CGroup of the bazel-built services should now contain something like:
    `/usr/bin/python3.8 /home/vagrant/magma/bazel-bin/orc8r/gateway/python/magma/magmad/magmad.runfiles/__main__/orc8r/gateway/python/magma/magmad/main.py` 
- To monitor the (bazel-built) services `journalctl` can be used.

#### Precommit integration tests
- How to run the (precommit) integration tests: see [documentation](https://docs.magmacore.org/docs/next/lte/s1ap_tests#s1ap-integration-tests ). 
- Note: It seems that the magma_trfserver vm needs to be running in order for all precommit integration tests to run successfully. 
- **All :100: precommit integration tests ran successfully on the bazel-built :seedling:  services in this setup** :bangbang:
- **Update:** All **sanity integration tests** ran successfully on the bazel-built services in this setup:bangbang:

### LINKED ISSUE #12057: Generate workflow for bazel built MME test of `lte integ-test`
Exit criteria: Run LTE integration tests with Bazel built MME

### PR DESCRIPTION
**Pull request by LKreutzer**
Signed-off-by: Lars Kreutzer <lars.kreutzer@tngtech.com>

## Summary

- Create LTE integ test workflow that runs the LTE integ tests on the bazel-built services.
  - Create service files for all services that start the bazel-built executables.
  - Service files are copied at runtime into the `/etc/systemd/system/` folder.
- The `.yml` workflow file is kept as close as possible to the Make LTE integ tests to make sure that the environment is the same.
  - The code duplication is temporary until the bazel switchover happens.
  - Adapting the workflow to both build systems adds too much complexity. 
- New fabric functions for running the integ tests are created 
  - This was necessary in order to be able to run the bazel build with `vagrant ssh` directly from the runner which avoids going through fabric
  - Fabric leads to a massive slowdown of the bazel build due to logging piling up
  - See also #12963 
- Note: The `/etc/gai.conf` file is temporarily - for the bazel build - modified in order to prefer IPv4 connections which speeds up the bazel build.
- During the (re-)start of the AGW usually another `make build` is executed, to avoid this a new fabric function that only does a `make restart` is created.
- Resolves https://github.com/magma/magma/issues/12057

## Test Plan

Manually run the workflow.
1. [Run on fork](https://github.com/LKreutzer/magma/actions/runs/2690050327)
2. [Run on fork](https://github.com/LKreutzer/magma/actions/runs/2696194463)
3. [Run on fork](https://github.com/LKreutzer/magma/actions/runs/2696193874)
4. [Run on fork](https://github.com/LKreutzer/magma/actions/runs/2703791422)
5. [Run on fork](https://github.com/LKreutzer/magma/actions/runs/2703793455)

## Additional Information
- Currently the bazel remote cache is not used in this workflow
  - This saves remote caching costs
  - Currently there is an issue when building inside the VM on a MacOs runner with the remote cache that leads to IO errors and failing builds 
- Note: We do not use a GH matrix workflow here in order to keep the complexity low, to not disrupt the result reporting from the standard LTE integ test workflow and because we expect further modifications of this workflow that may be incompatible with the standard workflow.
- See also https://github.com/magma/magma/issues/13353

- [ ] This change is backwards-breaking

### CODE REVIEW

**Review by jheidbrink — APPROVED** [review]

File: .github/workflows/lte-integ-test-bazel.yml [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+    branches:
+      - master
+      - 'v1.*'
+    types:
+      - completed
+
+jobs:
+  lte-integ-test-bazel:
+    if: github.repository_owner == 'magma' || github.event_name == 'workflow_dispatch'
+    runs-on: macos-10.15
```
Comment (Neudrino) [inline .github/workflows/lte-integ-test-bazel.yml:19]: If we introduce a new test here, I think we should directly make it compatible with the newer macOS runners of at least version 11, as the [version 10.15 runners are becoming deprecated](https://github.com/magma/magma/issues/13337) soon.

File: .github/workflows/lte-integ-test-bazel.yml [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+    env:
+      SHA: ${{ github.event.workflow_run.head_commit.id || github.sha }}
+    steps:
+      - uses: actions/checkout@7884fcad6b5d53d10323aee724dc68d8b9096a2e # pin@v2
+        with:
+          ref: ${{ env.SHA }}
```
Comment (Neudrino) [inline .github/workflows/lte-integ-test-bazel.yml:25]: Why is this needed? Is that not the default behaviour to check-out the SHA of the triggering event?

File: .github/workflows/lte-integ-test-bazel.yml [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+      - completed
+
+jobs:
+  lte-integ-test-bazel:
+    if: github.repository_owner == 'magma' || github.event_name == 'workflow_dispatch'
+    runs-on: macos-10.15
+    env:
+      SHA: ${{ github.event.workflow_run.head_commit.id || github.sha }}
+    steps:
+      - uses: actions/checkout@7884fcad6b5d53d10323aee724dc68d8b9096a2e # pin@v2
```
Comment (Neudrino) [inline .github/workflows/lte-integ-test-bazel.yml:23]: We might want to use the up to date v3 for the new workflow right away, here.

File: .github/workflows/lte-integ-test-bazel.yml [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+        with:
+          ref: ${{ env.SHA }}
+      - name: Cache magma-dev-box
+        uses: actions/cache@v3
```
Comment (Neudrino) [inline .github/workflows/lte-integ-test-bazel.yml]: Can we also pin this on a hash, as for the checkout action for example, please.

File: .github/workflows/lte-integ-test-bazel.yml [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+        with:
+          path: ~/.vagrant.d/boxes/magmacore-VAGRANTSLASH-magma_trfserver
+          key: vagrant-box-magma-trfserver
+      - name: setup pyenv
+        uses: "gabrielfalcao/pyenv-action@5327db2939908b2ef8f62d284403d678c4b611d0" # pin@v8
+        with:
+          default: 3.8.5
+      - uses: actions/setup-python@7f80679172b057fc5e90d70d197929d454754a5a # pin@v2
+        with:
+          python-version: '3.8.10'
```
Comment (Neudrino) [inline .github/workflows/lte-integ-test-bazel.yml:47]: This is confusing. Why are we installing two different minor versions of Python in two different ways here?

File: .github/workflows/lte-integ-test-bazel.yml [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+        run: |
+          pip3 install --upgrade pip
+          pip3 install ansible fabric3 jsonpickle requests PyYAML firebase_admin
+          vagrant plugin install vagrant-vbguest vagrant-disksize
+      - name: Open up network interfaces for VM
+        run: |
+          sudo mkdir -p /etc/vbox/
+          sudo touch /etc/vbox/networks.conf
+          sudo sh -c "echo '* 192.168.0.0/16' > /etc/vbox/networks.conf"
+          sudo sh -c "echo '* 3001::/64' >> /etc/vbox/networks.conf"
```
Comment (Neudrino) [inline .github/workflows/lte-integ-test-bazel.yml:58]: - The usual workaround is something like `echo "bla" | sudo tee $file`. Is `tee` not available on this runner?
- Should we not append also the first entry, in case there is already something configured? (At least thats what the quik start guide does.)

File: .github/workflows/lte-integ-test-bazel.yml [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+          vagrant ssh -c 'cd ~/magma; bazel build --profile=bazel_profile_lte_integ_tests `bazel query "kind(.*_binary, //orc8r/... union //lte/...)"`;' magma
+          vagrant ssh -c 'sudo sed -i "s@precedence ::ffff:0:0/96  100@#precedence ::ffff:0:0/96  100@" /etc/gai.conf;' magma
+      - name: Run the integ test
+        run: |
+          cd lte/gateway
+          export MAGMA_DEV_CPUS=3
+          export MAGMA_DEV_MEMORY_MB=9216
+          fab bazel_integ_test_post_build
+      - name: Upload bazel profile
+        uses: actions/upload-artifact@v3
```
Comment (Neudrino) [inline .github/workflows/lte-integ-test-bazel.yml]: We should pin the version on a hash here as well.

File: .github/workflows/lte-integ-test-bazel.yml [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+        if: always()
+        with:
+          name: Bazel profile lte integ tests
+          path: bazel_profile_lte_integ_tests
+      - name: Get test results
+        if: always()
+        run: |
+          cd lte/gateway
+          fab get_test_summaries:dst_path="test-results"
+          ls -R
```
Comment (Neudrino) [inline .github/workflows/lte-integ-test-bazel.yml:88]: This looks like for debugging. Is it? (In case it should be removed.)

File: .github/workflows/lte-integ-test-bazel.yml (hunk 7 above)
Comment (LKreutzer) [reply .github/workflows/lte-integ-test-bazel.yml]: Done

File: .github/workflows/lte-integ-test-bazel.yml (hunk 1 above)
Comment (LKreutzer) [reply .github/workflows/lte-integ-test-bazel.yml:19]: See updated PR description:
- The `.yml` workflow file is kept as close as possible to the Make LTE integ tests to make sure that the environment is the same.
  - The code duplication is temporary until the bazel switchover happens.
  - Adapting the workflow to both build systems adds too much complexity.

File: .github/workflows/lte-integ-test-bazel.yml (hunk 2 above)
Comment (LKreutzer) [reply .github/workflows/lte-integ-test-bazel.yml:25]: See updated PR description

File: .github/workflows/lte-integ-test-bazel.yml (hunk 3 above)
Comment (LKreutzer) [reply .github/workflows/lte-integ-test-bazel.yml:23]: See updated PR description

File: .github/workflows/lte-integ-test-bazel.yml (hunk 5 above)
Comment (LKreutzer) [reply .github/workflows/lte-integ-test-bazel.yml:47]: See updated PR description

File: .github/workflows/lte-integ-test-bazel.yml (hunk 6 above)
Comment (LKreutzer) [reply .github/workflows/lte-integ-test-bazel.yml:58]: See updated PR description

File: .github/workflows/lte-integ-test-bazel.yml (hunk 8 above)
Comment (LKreutzer) [reply .github/workflows/lte-integ-test-bazel.yml:88]: See updated PR description

File: .github/workflows/lte-integ-test-bazel.yml (hunk 4 above)
Comment (LKreutzer) [reply .github/workflows/lte-integ-test-bazel.yml]: I pinned the action on: `actions/cache@0865c47f36e68161719c5b124609996bb5c40129 # pin@v3` and created a PR to do the same for the workflows that are already on master: https://github.com/magma/magma/pull/13348

**Review by Neudrino — APPROVED** [review]
> See updated PR description

We should create a ticket to clean up the unattended comments here to get to a clean state at some point with this newly added functionality than.

**Review by sebathomas — APPROVED** [review]
